stratosphere/loader/source/ldr_meta.cpp

```diff
@@ -0,0 +1,217 @@
+
+        Result ValidateAci(const Aci *aci, size_t size) {
+            /* Validate magic. */
+            if (aci->magic != Aci::Magic) {
+                return ResultLoaderInvalidMeta;
+            }
+
+            /* Validate Fah, Sac, Kac. */
+            R_TRY(ValidateSubregion(sizeof(Aci), size, aci->fah_offset, aci->fah_size));
+            R_TRY(ValidateSubregion(sizeof(Aci), size, aci->sac_offset, aci->sac_size));
+            R_TRY(ValidateSubregion(sizeof(Aci), size, aci->kac_offset, aci->kac_size));
+
+            return ResultSuccess;
+        }
+
+        Result LoadMetaFromFile(FILE *f, MetaCache *cache) {
+            /* Reset cache. */
+            cache->meta = {};
+
+            /* Read from file. */
+            size_t npdm_size = 0;
+            {
+                /* Get file size. */
+                fseek(f, 0, SEEK_END);
+                npdm_size = ftell(f);
+                fseek(f, 0, SEEK_SET);
+
+                /* Read data into cache buffer. */
+                if (npdm_size > MetaCacheBufferSize || fread(cache->buffer, npdm_size, 1, f) != 1) {
+                    return ResultLoaderTooLargeMeta;
+                }
+            }
+
+            /* Ensure size is big enough. */
+            if (npdm_size < sizeof(Npdm)) {
+                return ResultLoaderInvalidMeta;
+            }
+
+            /* Validate the meta. */
+            {
+                Meta *meta = &cache->meta;
+
+                Npdm *npdm = reinterpret_cast<Npdm *>(cache->buffer);
+                R_TRY(ValidateNpdm(npdm, npdm_size));
+
+                Acid *acid = reinterpret_cast<Acid *>(cache->buffer + npdm->acid_offset);
+                Aci *aci = reinterpret_cast<Aci *>(cache->buffer + npdm->aci_offset);
+                R_TRY(ValidateAcid(acid, npdm->acid_size));
+                R_TRY(ValidateAci(aci, npdm->aci_size));
+
+                /* Set Meta members. */
+                meta->npdm = npdm;
+                meta->acid = acid;
+                meta->aci = aci;
+
+                meta->acid_fac = reinterpret_cast<u8 *>(acid) + acid->fac_offset;
+                meta->acid_sac = reinterpret_cast<u8 *>(acid) + acid->sac_offset;
+                meta->acid_kac = reinterpret_cast<u8 *>(acid) + acid->kac_offset;
+
+                meta->aci_fah = reinterpret_cast<u8 *>(aci) + aci->fah_offset;
+                meta->aci_sac = reinterpret_cast<u8 *>(aci) + aci->sac_offset;
+                meta->aci_kac = reinterpret_cast<u8 *>(aci) + aci->kac_offset;
+            }
+
+            return ResultSuccess;
+        }
+
+    }
+
+    /* API. */
+    Result LoadMeta(Meta *out_meta, ncm::TitleId title_id) {
+        FILE *f = nullptr;
+
+        /* Try to load meta from file. */
+        R_TRY(OpenCodeFile(f, title_id, MetaFilePath));
+        {
+            ON_SCOPE_EXIT { fclose(f); };
+            R_TRY(LoadMetaFromFile(f, &g_meta_cache));
+        }
+
+        /* Patch meta. Start by setting all title ids to the current title id. */
```

Comment:
In theory the worst case is that this is a NOP since official titles should already have these values set to what Atmosphère forces them to, and it also makes HBL support work.

Reply:
This is necessary to not break users who are loading HBL as an explicit NSP title override instead of properly, so you can consider it a kindness.